File: Data Structures & Algorithms/permutations/submission-3.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        n = len(nums)
        res = []

        def util(i):
            if i == n:
                res.append(nums[:])
                return

            for j in range(i, n):
                nums[j], nums[i] = nums[i], nums[j]
                util(i+1)
                nums[j], nums[i] = nums[i], nums[j]

        util(0)
        return res

File: Data Structures & Algorithms/burst-balloons/submission-3.cpp
class Solution {
public:
    vector<vector<int>> dp;
    int maxCoins(vector<int>& nums) {
        int n = nums.size();
        dp.assign(n+2, vector<int> (n+2, 0));
        nums.insert(nums.begin(), 1);
        nums.push_back(1);
        return util(1, n, nums);
    }

    int util(int l, int r, vector<int>& nums) {
        if (dp[l][r] != 0) return dp[l][r];
        int ans = 0;
        for (int i = l; i <= r; i++) {
            int curr = (nums[i] * nums[l-1] * nums[r+1]) + util(l, i-1, nums) + util(i+1, r, nums);
            ans = max(ans, curr);
        }
        return dp[l][r] = ans;
    }
};
